src/optimizer/plan_nodes/logical_filter.rs

```diff
@@ -52,10 +52,116 @@ impl PlanNode for LogicalFilter {
     fn estimated_cardinality(&self) -> usize {
         self.child().estimated_cardinality()
     }
+
+    fn prune_col(&self, required_cols: BitSet) -> PlanRef {
+        struct CollectRequiredCols(BitSet);
+        impl ExprVisitor for CollectRequiredCols {
+            fn visit_input_ref(&mut self, expr: &BoundInputRef) {
+                self.0.insert(expr.index);
+            }
+        }
+        let mut visitor = CollectRequiredCols(required_cols);
+        visitor.visit_expr(&self.expr);
+
+        struct ShiftLeft(usize);
+        impl ExprRewriter for ShiftLeft {
+            fn rewrite_input_ref(&self, expr: &mut BoundExpr) {
+                match expr {
+                    BoundExpr::InputRef(ref mut input_ref) => {
+                        input_ref.index -= self.0;
+                    }
+                    _ => unreachable!(),
+                }
+            }
+        }
```

Comment:
Shift is not enough: input refs `[1,3]` should become `[0,1]`. But since we don't have sub-queries like `select v1, v2 from (select v1,v3 from t)`, this is hard to trigger. 🤪